File: src/main/java/no/kvikshaug/gh/modules/IMDb.java
package no.kvikshaug.gh.modules;

import no.kvikshaug.gh.Grouphug;
import no.kvikshaug.gh.ModuleHandler;
import no.kvikshaug.gh.exceptions.NoTitleException;
import no.kvikshaug.gh.listeners.TriggerListener;
import no.kvikshaug.gh.util.Web;

import java.io.BufferedReader;
import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URL;

import org.apache.commons.io.IOUtils;
import org.jdom.Document;
import org.jdom.Element;
import org.jdom.JDOMException;
import org.jdom.xpath.XPath;

public class IMDb implements TriggerListener {

	private static final String TRIGGER = "imdb";
	private static final String TRIGGER_HELP = "imdb";

	public IMDb(ModuleHandler moduleHandler) {
		moduleHandler.addTriggerListener(TRIGGER, this);
		moduleHandler.registerHelp(TRIGGER_HELP, "IMDb: Show IMDb info for a movie\n" +
				"  "+Grouphug.MAIN_TRIGGER+TRIGGER +"<movie name>");
		System.out.println("IMDb module loaded.");
	}


	public void onTrigger(String channel, String sender, String login, String hostname, String message, String trigger) {
		URL imdbURL;
		try {
			imdbURL = Web.googleSearch(message+"+site:www.imdb.com").get(0);
		} catch(IndexOutOfBoundsException ex) {
			Grouphug.getInstance().sendMessage("Sorry, I didn't find "+message+" on IMDb.");
			return;
		} catch(IOException e) {
			Grouphug.getInstance().sendMessage("But I don't want to. (IOException)");
			return;
		} catch (JDOMException e) {
			Grouphug.getInstance().sendMessage("I seem to have thrown a JDOMException. Woopsie!");
			return;
		}


		String title = "";
		String score = "";
		String plot = "";

		Document doc;
		try {
			doc = Web.getJDOMDocument(imdbURL);

			title = Web.fetchTitle(imdbURL);
			title = title.substring(0, title.length()-7); //To remove the  - IMDb part of the title

			// find the score of the movie element using XPath
			XPath scorePath = XPath.newInstance("//h:span[@id='star-bar-user-rate']/h:b");
			scorePath.addNamespace("h","http://www.w3.org/1999/xhtml");

			Element scoreElement = (Element)scorePath.selectSingleNode(doc);
			if(scoreElement == null) {
				throw new JDOMException("No score element in DOM");
			}
			score = scoreElement.getText();


			// find the score of the movie element using XPath
			XPath plotPath = XPath.newInstance("//h:table[@id='title-overview-widget-layout']/h:tbody/h:tr/h:td[@id='overview-top']/h:p[2]");
			plotPath.addNamespace("h","http://www.w3.org/1999/xhtml");

			Element plotElement = (Element)plotPath.selectSingleNode(doc);
			if(plotElement == null) {
				throw new JDOMException("No plot element in DOM");
			}
			plot = plotElement.getText();


		} catch (IOException e) {
			e.printStackTrace();
			System.out.println(e.getMessage());
		} catch (JDOMException e) {
			e.printStackTrace();
			System.out.println(e.getMessage());
		} catch (NoTitleException e) {
			e.printStackTrace();
			System.out.println(e.getMessage());
		}
		try {
			Grouphug.getInstance().sendMessage(title+"\n"+plot+"\n"+"\n"+score+"/10\n"+imdbURL.toString());
		} catch(NullPointerException ex) {
			Grouphug.getInstance().sendMessage("The IMDb site layout may have changed, I was unable to parse it.");
		}
	}
}

File: src/main/scala/no/kvikshaug/gh/modules/IMDb.scala
package no.kvikshaug.gh.modules

import scala.xml._

import java.net.URL
import java.util.regex._
import org.ccil.cowan.tagsoup.jaxp.SAXParserImpl

import no.kvikshaug.gh.listeners.TriggerListener
import no.kvikshaug.gh.util.Web
import no.kvikshaug.gh.{Grouphug, ModuleHandler}

class IMDb(val handler: ModuleHandler) extends TriggerListener {

  val bot = Grouphug.getInstance
  handler.addTriggerListener("imdb", this)
  handler.registerHelp("imdb", "!imdb - Show IMDb info for a movie")

  def onTrigger(channel: String, sender: String, login: String, hostname: String, message: String, trigger: String): Unit = {
    try {
      val imdbURLs = Web.googleSearch(message+"+site:www.imdb.com")
      if(imdbURLs isEmpty) {
        bot.msg(channel, "Sorry, I didn't find "+message+" on IMDb.")
        return
      }
      val imdbURL = imdbURLs.get(0)
      if(!imdbURL.toString.contains("/title/tt")) {
        bot.msg(channel, "I found that on IMDb but it doesn't look like a movie: " + imdbURL)
        return
      }
      var title = Web.fetchTitle(imdbURL)
      title = title.substring(0, title.length()-7) //To remove the  - IMDb part of the title      
      val root = XML.withSAXParser(SAXParserImpl.newInstance(null)).load(
        Web.prepareEncodedBufferedReader(imdbURL))
      var score = ""
      for(span <- root \\ "span" if((span \ "@itemprop").text == "ratingValue")) {
        score = span.text + "/10"
      }
      var plot = ""
      for(p <- root \\ "p" if((p \ "@itemprop").text == "description")) {
        if(p.text.contains("See full summary")) {
          plot = p.text.substring(0, p.text.indexOf("See full summary")).trim
        } else {
          plot = p.text.trim          
        }
      }
      bot.msg(channel, title+"\n"+plot+"\n"+score+"\n"+imdbURL)
    } catch {
      case e => bot.msg(channel, "Meh, I got some exception, did IMDb change their layout AGAIN!?")
                e.printStackTrace
    }
  }
}

